File: Leetcode_solutions/Two-Sum.js
/**
 * @param {number[]} nums
 * @param {number} target
 * @return {number[]}
 */
var twoSum = function(nums, target) {
  combination = [];
  for (let i = 0; i < nums.length; i++) {
    for (let j = i + 1; j < nums.length; j++) {
      if (nums[i] + nums[j] === target) {
        combination.push(i, j);
      }
    }
  }
  return combination;
};

/**
 * @param {number[]} nums
 * @param {number} target
 * @return {number[]}
 */
var twoSum = function(nums, target) {
  store = {};
  for (let i = 0; i < nums.length; i++) {
    m = target - nums[i];
    if (m in store) {
      return [i, store[m]];
    } else {
      store[nums[i]] = i;
    }
  }
  return null;
};

console.log(twoSum([2, 7, 11, 15], 9));
console.log(twoSum([3, 2, 4], 6));

File: Leetcode_solutions/two-sum.py
class Solution:
    def twoSum(self, nums: List[int], target: int) -> List[int]:
        ans = []
        for i in range(len(nums)):
            for j in range(i + 1, len(nums)):
                if nums[i] + nums[j] == target:
                    ans = [i, j]
                    break
        return ans


class Solution:
    def twoSum(self, nums: List[int], target: int) -> List[int]:
        q = nums[:]
        while True:
            now = q[0]
            index = nums.index(now)
            q.pop(0)
            for i in range(index + 1, len(nums)):
                if now + nums[i] == target:
                    return [index, i]


class Solution:
    def twoSum(self, nums: List[int], target: int) -> List[int]:
        dict = collections.defaultdict(list)
        for i in range(len(nums)):
            complement = target - nums[i]
            if complement in dict.keys():
                return [dict[complement][0], i]
            dict[nums[i]].append(i)

from collections import defaultdict
class Solution:
    def twoSum(self, nums: List[int], target: int) -> List[int]:
        map = defaultdict(int)
        for i in range(len(nums)):
            map[nums[i]] = i
        for i in range(len(nums)):
            complement = target - nums[i]
            if map[complement] and map[complement] != i:
                return [i, map[complement]]

from collections import defaultdict
class Solution:
    def twoSum(self, nums: List[int], target: int) -> List[int]:
        map = defaultdict(int)        
        for i in range(len(nums)):
            complement = target - nums[i]
            if str(complement) in map and map.get(str(complement)) != i:
                return [i, map[str(complement)]]
            map[str(nums[i])]=i